### DISCUSSION TITLE
Center grid vertically

### DISCUSSION THREAD

**Discussion by Heloise-Truong** [tailwindlabs/tailwindcss#8633] [Help] [answered]
Hello, I have a grid with two rows which I would like to center vertically.
For the moment I have this : 
[  ]  [  ]  [  ]  [  ]  [  ]  [  ] 
[  ]  [  ] 
And I wish to have this : 
```
[  ]  [  ]  [  ]  [  ]  [  ]  [  ] 
            [  ]  [  ] 
```

Is it possible to do so ?

Here is my code : 
`            questionTxt.className = "grid grid-flow-row grid-cols-6 md:grid-rows-1 md:grid-flow-col lg:grid-rows-1 lg:grid-flow-col gap-x-1;";
`

Thank you in advance

**Comment by cannap** [comment]
@Heloise-Truong 

use code tags 
`

```
[ ] [ ] [ ] [ ] [ ] [ ]
[ ]                 [ ]
```

**Reply by Heloise-Truong** [reply]
Thank you very much, it's better now indeed

**Comment by rohmann** [comment, marked as answer]
Hi @Heloise-Truong 

Here's a link with a few examples: https://play.tailwindcss.com/0XwSniFU6V

If you're using a grid, each cell flows into the next available track by default. The grid "flow" will let you use a "dense" mode, but that is mainly for cases where you have cells spanning multiple tracks and you want to let smaller items fill in available space.

If you add `col-start-3` to your seventh "cell" in the grid, it will offset it by two positions. I wouldn't consider this actually "centering" them. They will be visually centered, but you're explicitly placing the items.

The layout you're looking for here where it automatically centers items **would work more naturally if you used flexbox**. Note that when `flex-basis` is a percentage, you can't effectively use gap. I did that example with a border instead. This can be worked around by subtracting the gap in CSS like `flex-basis: calc( 100% / 6 - 1px ); where `1px is your gap size and `6` is the number of columns, but getting that setup kinda takes away from the more general ideas I was trying to demonstrate here.

**Reply by Heloise-Truong** [reply]
Hello @rohmann ,
Thank you very much for your complete answer. I totally understand what you want to demonstrate and it looks like a good solution. However I can't seem to apply it to my code as my grid is dynamic : If I have more than 6 columns, then I put a second row.
Here is a quick explanation of my code : I have a string, and I put each char of the string in a box (thus my grid). If my string has more than 6 characters (so if first row has more than 6 columns), then I put a second row.
With your answer I unfortunately still don't know how to do it dynamically.

Here is my code : 
`            questionTxt.className = "grid grid-flow-row grid-cols-6 lg:grid-flow-col gap-x-1;";`
`            const html = [];`
`            //Parse string to create array of chars `
`            var questionParse = JSON.parse(currentQuestion.question);`
`            //Display letters in random order`
`            for (let i = questionParse.length -1; i >= 0; i--) {`
`                const randLetters = Math.floor(Math.random() * (i+1));//variable to have a random letter`
`                [questionParse[i], questionParse[randLetters]] = [questionParse[randLetters], questionParse[i]];`
`                //push the letter in a div balise with tailwind class`
`                html.push(``<div class="uppercase text-sm md:text-lg lg:text-2xl mx-0.3 mt-0.5 px-3 md:px-5 lg:px-5 border border-black rounded">${questionParse[i]}</div>\``);`
`            }`
`            questionTxt.innerHTML = html.join(" ");//display result`

Is it still possible to apply your answer with my speficiation ?

Thank you in advance

**Reply by rohmann** [reply]
I see, the additional context helps. What are you expecting it to do with an odd number of characters? With a grid setup where each cell is a full track, it will never be able to center properly one the second line if you have an odd number like nine characters.

Considering that, maybe you need to use 12 columns and have each cell span 2 tracks. Here's an example of that: https://play.tailwindcss.com/58S0u1F1Qd

There are a whole lot of different ways you can solve it dynamically, so I'll keep my advice specific to the actual classes needed to achieve the affect. I would suggest writing logic to find the 7th cell (if more than 6) and providing a class like `col-start-4` to inset the next row. That `col-start-*` class might need to be different depending on your number of letters.

And remember that Tailwind needs to see the full class, so you can't do `const class = `col-start-${number}`. It needs to be something like:

```
function resolveColStart(number) {
  if (number === 7) return 'col-start-4'; // note the full class name is included in your code
  // etc
}
```

**Reply by Heloise-Truong** [reply]
Thank you very much for your answer. It works now, but unfortunately it doesn't show ideally a smooth result 
ex: If I make the second row start from col2 then I can have those type of results : 
```
[  ]  [  ]  [  ]  [  ]  [  ]  [  ]  [  ]  
      [  ] 
```
or 
```
[  ]  [  ]  [  ]  [  ]  [  ]  [  ]  [  ]  
      [  ]  [  ]  [  ]  [  ]  [  ]  [  ] 
[  ] 
```
It would have been great if there was just a way to center grid vertically but it seems to not exist. Thank you very much for all the answers and possibilities you showed me. I think I will take the flex option so it can go to second row when the word is too big for phone, thank you

**Reply by rohmann** [reply]
Sounds good. Because of how CSS Grid is designed, everything must fall in alignment with the grid tracks. There are creative things you can do like using a higher resolution grid (more, smaller tracks) and spreading your cells across them but it's not really meant for fluid centering on additional rows. That's the perfect use case for flexbox though. Wishing you the best with the project.

**Reply by Heloise-Truong** [reply]
Thank you very much for your help !
